File: rust/crates/aoc_2021/scratch/day18.ts
// just scratch work for day 18
// noinspection JSUnusedGlobalSymbols

const LEFT_BRACE = -1;
const RIGHT_BRACE = -2;
const COMMA = -3;

function parseLine(input: string): number[] {
    const line: number[] = [];
    const a = input.split("");
    for (const c of a) {
        if (c === "[") {
            line.push(LEFT_BRACE);
            continue;
        }

        if (c === "]") {
            line.push(RIGHT_BRACE);
            continue;
        }

        if (c === ",") {
            line.push(COMMA);
            continue;
        }

        const n = parseInt(c, 10);
        line.push(n);
    }

    return line;
}

function add(a: number[], b: number[]): number[] {
    const newLine: number[] = [];
    newLine.push(LEFT_BRACE);
    newLine.push(...a);
    newLine.push(COMMA);
    newLine.push(...b);
    newLine.push(RIGHT_BRACE);
    return newLine;
}

function explode(line: number[]): boolean {
    let b: number = 0;
    for (let i = 0; i < line.length; i++) {
        if (line[i] === LEFT_BRACE) {
            b++;
        }

        if (line[i] === RIGHT_BRACE) {
            b--;
        }

        // Must be a number
        // Note that outer braces don't count
        if (b < 5) {
            continue;
        }

        // Must be a pair
        if (line[i] === COMMA && line[i - 1] >= 0 && line[i + 1] >= 0) {
            const beforeNum = line[i - 1];
            const afterNum = line[i + 1];
            for (let j = i - 2; j >= 0; j--) {
                if (line[j] < 0) {
                    continue;
                }

                line[j] += beforeNum;
                break;
            }

            for (let j = i + 2; j < line.length; j++) {
                if (line[j] < 0) {
                    continue;
                }

                line[j] += afterNum;
                break;
            }

            line.splice(i - 2, 5, 0);
            return true;
        }
    }

    return false;
}

function split(line: number[]): boolean {
    for (let i = 0; i < line.length; i++) {
        if (line[i] === LEFT_BRACE || line[i] === RIGHT_BRACE || line[i] === COMMA) {
            continue;
        }

        if (line[i] < 10) {
            continue;
        }

        const left = Math.floor(line[i] / 2);
        const right = line[i] - left;
        line.splice(i, 1);
        line.splice(i, 0, LEFT_BRACE, left, COMMA, right, RIGHT_BRACE);
        return true;
    }

    return false;
}

function getStringRepresentation(line: number[]): string {
    const s: string[] = [];
    for (const c of line) {
        switch (c) {
            case LEFT_BRACE: {
                s.push("[");
                break;
            }
            case RIGHT_BRACE: {
                s.push("]");
                break;
            }
            case COMMA: {
                s.push(",");
                break;
            }
            default: {
                s.push(c.toString());
                break;
            }
        }
    }

    return s.join("");
}

function calculateMagnitude(line: number[]): number {
    while (flatten(line)) { }
    return line[0];
}

function flatten(line: number[]): boolean {
    for (let i = 0; i < line.length; i++) {
        if (line[i] === COMMA && line[i - 1] >= 0 && line[i + 1] >= 0) {
            const left = line[i - 1];
            const right = line[i + 1];
            line.splice(i - 2, 5, 3 * left + 2 * right);
            return true;
        }
    }

    return false;
}

function run(problems: string[]): number {
    let start = parseLine(problems[0]);
    processLine(start);

    for (let i = 1; i < problems.length; i++) {
        start = add(start, parseLine(problems[i]));
        processLine(start);
    }

    return calculateMagnitude(start);
}

function processLine(line: number[]): void {
    while (true) {
        if (explode(line)) {
            continue;
        }

        if (split(line)) {
            continue;
        }

        break;
    }
}

function getHighestMagnitude(problems: string[]): number {
    let max = 0;
    for (let i = 0; i < problems.length; i++) {
        for (let j = 0; j < problems.length; j++) {
            if (i === j) {
                continue;
            }

            const a = run([problems[i], problems[j]]);
            const b = run([problems[j], problems[i]]);

            if (a > max) {
                max = a;
            }

            if (b > max) {
                max = b;
            }
        }
    }

    return max;
}

File: crates/aoc_2021/src/aoc/day18.rs
use common::day::{AoCProblem, Solution};
use std::cmp::max;

// Remark: At a later time, I'll try to implement this using a binary tree (which would have been
// ideal for a problem like this, except it was late and I was -- and still am -- lazy).

const LEFT_BRACE: i32 = -1;
const RIGHT_BRACE: i32 = -2;
const COMMA: i32 = -3;

pub struct Day18 {
    homework: Vec<String>,
}

// https://adventofcode.com/2021/day/18
impl AoCProblem for Day18 {
    fn prepare(input: String) -> Self {
        Self {
            homework: input.lines().map(|x| x.to_string()).collect(),
        }
    }

    fn part1(&mut self) -> Solution {
        let hw: Vec<&str> = self.homework.iter().map(|x| x.as_str()).collect();
        solve_homework_problem(&hw).into()
    }

    fn part2(&mut self) -> Solution {
        let hw: Vec<&str> = self.homework.iter().map(|x| x.as_str()).collect();
        let mut max_mag: i32 = i32::MIN;
        for i in 0..hw.len() {
            for j in 0..hw.len() {
                if i == j {
                    continue;
                }

                max_mag = max(
                    max(
                        solve_homework_problem(&[hw[i], hw[j]]),
                        solve_homework_problem(&[hw[j], hw[i]]),
                    ),
                    max_mag,
                );
            }
        }

        max_mag.into()
    }
}

/// Parses a snail number line.
///
/// # Parameters
/// - `line`: The line to parse.
///
/// # Returns
/// A vector that represents the snail number.
fn parse_line(line: &str) -> Vec<i32> {
    line.chars()
        .map(|x| match x {
            '[' => LEFT_BRACE,
            ']' => RIGHT_BRACE,
            ',' => COMMA,
            _ => x.to_digit(10).unwrap() as i32,
        })
        .collect()
}

/// Adds two snail numbers together.
///
/// # Parameters
/// - `a`: The first snail number.
/// - `b`: The second snail number.
///
/// # Returns
/// A vector containing the new snail number.
fn add(a: &[i32], b: &[i32]) -> Vec<i32> {
    // a + b = [(all of a), (all of b)]
    let mut v: Vec<i32> = vec![LEFT_BRACE];
    v.extend(a);
    v.push(COMMA);
    v.extend(b);
    v.push(RIGHT_BRACE);
    v
}

/// Attempts to explode a snail number.
///
/// # Parameters
/// - `n`: The snail number.
///
/// # Returns
/// `true` if something changed; `false` otherwise.
fn explode(n: &mut Vec<i32>) -> bool {
    let mut b = 0;
    for i in 0..n.len() {
        if n[i] == LEFT_BRACE {
            b += 1;
            continue;
        }

        if n[i] == RIGHT_BRACE {
            b -= 1;
            continue;
        }

        // We use 5 because we also count the outer-most braces.
        if b < 5 {
            continue;
        }

        // If this is a valid pair...
        if n[i] == COMMA && n[i - 1] >= 0 && n[i + 1] >= 0 {
            let before_num = n[i - 1];
            let after_num = n[i + 1];
            // Look for the nearest valid left number to add to
            for j in (0..(i - 2)).rev() {
                if n[j] < 0 {
                    continue;
                }

                n[j] += before_num;
                break;
            }

            // Look for the nearest valid right number to add to
            for val in n.iter_mut().skip(i + 2) {
                if *val < 0 {
                    continue;
                }

                *val += after_num;
                break;
            }

            // Index i looks like:
            //      ...[z,[x,y]],k...
            //              ^ i
            // So, we remove everything from (i - 2), inclusive, to (i + 3), exclusive and then
            // put a singular 0 in its place:
            //      ...[z,0],k...
            //            ^ i
            n.splice((i - 2)..(i + 3), [0]);
            return true;
        }
    }

    false
}

/// Attempts to split a snail number.
///
/// # Parameters
/// - `n`: The snail number.
///
/// # Returns
/// `true` if something changed; `false` otherwise.
fn split(n: &mut Vec<i32>) -> bool {
    for i in 0..n.len() {
        match n[i] {
            LEFT_BRACE | RIGHT_BRACE | COMMA => continue,
            _ if n[i] < 10 => continue,
            _ => {
                let left = n[i] / 2;
                let right = n[i] - left;
                n.splice(i..(i + 1), [LEFT_BRACE, left, COMMA, right, RIGHT_BRACE]);
                return true;
            }
        };
    }

    false
}

/// Gets the string representation of this snail number. Useful for debugging.
///
/// # Parameters
/// - `n`: The snail number.
///
/// # Returns
/// The string representation of this snail number.
fn get_string_representation(n: &[i32]) -> String {
    let mut s: String = String::new();
    for c in n {
        match *c {
            LEFT_BRACE => s.push('['),
            RIGHT_BRACE => s.push(']'),
            COMMA => s.push(','),
            _ => s.push_str(&c.to_string()),
        };
    }

    s
}

/// Flattens the snail number, essentially "simplifying" it.
///
/// # Parameters
/// - `n`: The snail number.
///
/// # Returns
/// `true` if the snail number was flattened once and `false` otherwise.
fn flatten(n: &mut Vec<i32>) -> bool {
    for i in 0..n.len() {
        if n[i] == COMMA && n[i - 1] >= 0 && n[i + 1] >= 0 {
            let left = n[i - 1];
            let right = n[i + 1];
            // Remove from index (i - 2) 5 times. Since index i looks like:
            //          [x,y]
            //            ^
            //            i
            // We can remove everything from (i - 2), inclusive, to (i + 3), exclusive and then
            // put the result of the operation in its place; in this case, 3 * x + 2 * y.
            n.splice((i - 2)..(i + 3), [3 * left + 2 * right]);
            return true;
        }
    }

    false
}

/// Solves a homework problem.
///
/// # Parameters
/// - `problems`: The addition problems to solve.
///
/// # Returns
/// The result (magnitude).
fn solve_homework_problem(problems: &[&str]) -> i32 {
    let mut start = parse_line(problems[0]);
    process_line(&mut start);

    for p in problems.iter().skip(1) {
        start = add(&start, &parse_line(p));
        process_line(&mut start);
    }

    while flatten(&mut start) {}
    start[0]
}

/// Continuously processes a line until it can no longer be processed. This will continuously run
/// the `split` and `explode` methods.
///
/// # Parameters
/// - `n`: The snail number to process.
fn process_line(n: &mut Vec<i32>) {
    loop {
        if explode(n) {
            continue;
        }

        if split(n) {
            continue;
        }

        break;
    }
}
